src/Microsoft.AspNetCore.Blazor/Components/BlazorComponent.cs

```diff
@@ -169,14 +169,26 @@ private void HandleException(Exception ex)
             Console.Error.WriteLine($"[{ex.GetType().FullName}] {ex.Message}\n{ex.StackTrace}");
         }
 
-        void IHandleEvent.HandleEvent(UIEventHandler handler, UIEventArgs args)
+        Task IHandleEvent.HandleEvent(Func<UIEventArgs, Task> handler, UIEventArgs args)
         {
-            handler(args);
+            var task = handler(args);
 
             // After each event, we synchronously re-render (unless !ShouldRender())
             // This just saves the developer the trouble of putting "StateHasChanged();"
             // at the end of every event callback.
             StateHasChanged();
+
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                return task;
+            }
+
+            return task.ContinueWith(OnAsyncTaskCompleted);
+        }
+
+        protected virtual void OnAsyncTaskCompleted(Task task)
```

Comment:
We can also turn this into another extension point `OnAsyncTaskFaulted(Task)` that would be called in the case of failures. I'm not really convinced that this is useful at all really. Because you could just try/catch in your own code.

I also resisted the urge to create and register these tasks in any sort of global queue. This means that you could have a situation where a task is running while a component is removed from the render tree. I'm not sure how sophisticated we want to get with that sort of thing.

Reply:
I agree that `OnAsyncTaskFaulted` would not be needed, and would make the same case about `OnAsyncTaskCompleted` (at least, that it should be private).

Reply:
Ah yes, I agree we most likely shouldn't control that at the framework level. If a component is removed from the render tree (and hence disposed) while an async operation is in progress, we don't know whether or not that means the task should be cancelled. If it's a "Save" operation, for example, it shouldn't be cancelled, whereas if it's a "Fetch some new data to display", it could be cancelled. It's up to the developer to put in their own logic if they want to cancel.

Note that by design it's not harmful to attempt rendering after disposal. The renderer will just ignore the call when it sees the component is already disposed.